### PR TITLE
Report the type of change observed in FileWatcher

### LINKED ISSUE #53: Make `FileWatcher` return the type of change observed
Right now the `FileWatcher` only returns the path of the changed file, but it doesn't say if a change was a creation, modification or deletion, which might be useful information to have.

_Originally posted by @shsms in https://github.com/frequenz-floss/frequenz-channels-python/issues/51#issuecomment-1320171672_

### PR DESCRIPTION
**Pull request by daniel-zullo-frequenz**
Previously, the FileWatcher only provided information on the path of the changed file, without indicating whether the change was a creation, modification, or deletion. This commit improves the functionality of the FileWatcher by adding the ability to report the type of change observed. This additional information can be useful when tracking changes to files and understanding the context of the change.

Fixes #53

### CODE REVIEW

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/channels/util/_file_watcher.py]: This should actually return a `FileWatcher.EventType`, the underlying library used to implement `FileWatcher` is hidden to the user.

I would also recommended using a `dataclass` as the return:

```python
@dataclass(frozen=True)
class Event:
    type: EventType
    path: pathlib.Path
```

It avoids errors and makes the code more readable and future proof (if we add more attributes users don't need to update the code (if they are using tuple unpacking syntax, adding a new item to the tuple will make `(change, path) = fw.receive()` fail).

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/channels/util/_file_watcher.py]: When you update old code to add more built-in types it would be nice if you convert it to using `from __future__ import annotations` instead. It is optional but very desirable so we can slowly get rid of the old idioms.

**Review by leandro-lucarella-frequenz — COMMENTED** [review]
I propose some improvements to the API.

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: this suggestion makes totally sense, I'll update it accordingly. Thanks!

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: The Tuple will no longer be needed as we will use the `Event` class, however I'll add a commit to use annotations from __future__  instead

**Comment by daniel-zullo-frequenz** [comment]
Updated based on a couple of suggestions from Luca.

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/channels/util/_file_watcher.py]: This still needs to be `FileWatcher.EventType`, not `Change`. We shouldn't expose anything from `watchfiles` to the user.

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [inline src/frequenz/channels/util/_file_watcher.py]: Any reason not to put `FileEvent` inside `FileWatcher`, just like `EventType`?

IIRC `EventType` was in the global scope too before, and I think it was changed as part of a restructuring to add the `util` package. I think I did this without thinking too much about google-style imports, and with that in mind it might not be the best approach. It is something we might need to review in the future. But for now I would try to keep the consistency.

Either both inside `FileWatcher` or both outside. About names, if it's inside, I think just plain `Event` is fine, because it must always be referred to by `FileWatcher.Event`. Even more, now the type could be nested in event: `FileWatcher.Event.Type`.

If it goes outside, I think `FileEvent` sounds good, then I would make `EventType` also global for consistency and call it `FileEventType`.

What I like about nesting is that is less import and the context of what each type is used for is super clear, but the downside is there might be more typing.

File: tests/utils/test_integration.py
Comment (leandro-lucarella-frequenz) [inline tests/utils/test_integration.py]: This should never be imported in integration tests, as we shouldn't expose `watchfiles` to the end user.

Comment (daniel-zullo-frequenz) [reply tests/utils/test_integration.py]: you're absolutely right. I should have used FileWatcher.EventType instead. Will update it

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: yeah my bad, will update it

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: > Any reason not to put FileEvent inside FileWatcher, just like EventType?

The reason was FileWatcher  is a `Receiver[FileEvent]` (inherits from). I would like to define Event internally if that's possible. I'll need to check it out

Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: Actually it works just quoting `FileWatcher.Event` in the class signature as follows:
```python
class FileWatcher(Receiver["FileWatcher.Event"]):
    ...
```

Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: Then I'll define `Event` inside `FileWatcher`

Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: done. Renamed `FileEvent` to `Event` and moved it inside `FileWatcher`

**Comment by daniel-zullo-frequenz** [comment]
Updated based on latest suggestions from Luca

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: Ok, interesting. It's surprising that it works as it is a class parametrized on itself 😱. I guess python can get away with a lot of crazy shit thanks to being so dynamic.

**Review by leandro-lucarella-frequenz — DISMISSED** [review]

**Comment by daniel-zullo-frequenz** [comment]
@leandro-lucarella-frequenz I added a trivial fix to correct typos and format in the releases notes (not related to this PR). Can you please check it and re-approve it if LGTY?

File: RELEASE_NOTES.md
Comment (leandro-lucarella-frequenz) [inline RELEASE_NOTES.md]: I started using different bullets types for different levels to make @shsms happy, I leave this to you, I don't care about either style :)

**Review by leandro-lucarella-frequenz — DISMISSED** [review]
Just one super minor comment left, but approving anyway because it is just style and for a file that will be cleaned up right after this is merged.

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: hmm MD linter complains about it with: `MD004/ul-style: Unordered list style [Expected: asterisk; Actual: dash]markdownlint[MD004](https://github.com/DavidAnson/markdownlint/blob/v0.28.2/doc/md004.md)`

Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: ah I should be replaced by `+` instead to make Sahas happy :), if there is another subitem then it should be `-`
I'll change that

Comment (leandro-lucarella-frequenz) [reply RELEASE_NOTES.md]: That's weird. It seems the default option is `consistent` and if I understand correctly this is consistent, or are there any other sub-lists using different bullets styles?

File: src/frequenz/channels/util/_file_watcher.py
Comment (Marenz) [inline src/frequenz/channels/util/_file_watcher.py]: I don't think events need to be notified :)
Maybe just "Class used to notify about file changes"?

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: The linter still complains :(
I'll revert the format changes

Comment (leandro-lucarella-frequenz) [reply RELEASE_NOTES.md]: Oh, no, it seems like with `sublist` option that would be valid. Not sure, as I said I don't care.  Whatever we decide it would be good to use that linter in the CI so we don't need to have these discussions anymore :D

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: Sounds good, will update it

File: RELEASE_NOTES.md
Comment (leandro-lucarella-frequenz) [reply RELEASE_NOTES.md]: I would leave it as it is now to avoid the controversy in this PR. @daniel-zullo-frequenz  can you create a **Discussion** for this? Maybe in https://github.com/frequenz-floss/frequenz-repo-config-python/discussions/categories/ideas, so we just plug the linter in there directly.

File: src/frequenz/channels/util/_file_watcher.py
Comment (leandro-lucarella-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: I don't get this, I don't read from there that events need to be modified. But now that you are moving my attention there, I think we should avoid starting class documentation with "A class to" and just say "A file change event." or something like that.

Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: done

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: o, i'll create a Discussion

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py]: I didn't see the last comment. @Marenz  does `A file change event.`  sound go to you?

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [reply RELEASE_NOTES.md]: Added here https://github.com/frequenz-floss/frequenz-repo-config-python/discussions/23

File: src/frequenz/channels/util/_file_watcher.py
Comment (Marenz) [inline src/frequenz/channels/util/_file_watcher.py:42]: Might it be better to remove the `None` part and just default-init it with `set(FileWatcher.EventType)`? Then we can remove the `if` in line `51`

Comment (leandro-lucarella-frequenz) [reply src/frequenz/channels/util/_file_watcher.py:42]: That's not a bad idea, but we would need to use `frozenset(FileWatcher.EventType)` then, because `set` is mutable and [the problem with mutable default arguments in Python](https://editorconfig.org/). If we use `frozenset()` for the default, then `event_types` type should be `collections.abc.Set` (or maybe `collections.abc.Iterable` would be enough if we convert it to a `frozenset` internally, then the argument type can be `Iterable` and the attribute type `frozenset`, so it can't be modified either via the attribute).

**Review by leandro-lucarella-frequenz — APPROVED** [review]
Besides agreeing to @Marenz suggestion, for me it is fine if it's left for a different PR so we can merge this sooner than later, as it is out of the scope of this PR.

File: src/frequenz/channels/util/_file_watcher.py
Comment (Marenz) [reply src/frequenz/channels/util/_file_watcher.py:42]: Good observation!

**Comment by daniel-zullo-frequenz** [comment]
> Besides agreeing to @Marenz suggestion, for me it is fine if it's left for a different PR so we can merge this sooner than later, as it is out of the scope of this PR.

Ok, I'll merge this PR as it currently is and create an issue to address https://github.com/frequenz-floss/frequenz-channels-python/pull/87#discussion_r1191102147

File: src/frequenz/channels/util/_file_watcher.py
Comment (daniel-zullo-frequenz) [reply src/frequenz/channels/util/_file_watcher.py:42]: created the issue https://github.com/frequenz-floss/frequenz-channels-python/issues/88

**Pull request merged**
